artifact/gem5art/artifact/_artifactdb.py

```diff
@@ -182,15 +193,49 @@ def searchByLikeNameType(self, name: str, typ: str, limit: int) -> Iterable[Dict
 
 _db = None
 
-def getDBConnection(typ: Type[ArtifactDB] = ArtifactMongoDB) -> ArtifactDB:
+_db_schemes : Dict[str, Type[ArtifactDB]] = {
+    'mongodb': ArtifactMongoDB
+}
+
+def _getDBType(uri: str) -> Type[ArtifactDB]:
+    """Internal function to take a URI and return a class that can be
+    constructed with that URI. For instance "mongodb://localhost" will return
+    an ArtifactMongoDB. More types will be added in the future.
+
+    Supported types:
+        **ArtifactMongoDB**: mongodb://...
+            See http://dochub.mongodb.org/core/connections for details.
+    """
+    result = urlparse(uri)
+    if result.scheme in _db_schemes:
+        return _db_schemes[result.scheme]
+    else:
+        raise Exception(f"Cannot find DB type for {uri}")
+
+def getDBConnection(uri: str = '') -> ArtifactDB:
     """Returns the database connection
 
-    Eventually, this should likely read from a config file to get the database
-    information. However, for now, we'll use mongodb defaults
+    uri: a string representing the URI of the database. See _getDBType for
+        details. If no URI is given we use the default
+        (mongodb://localhost:27017) or the value in the GEM5ART_DB environment
+        variable.
+
+    If the connection has not been established, this will create a new
+    connection.
     """
     global _db
 
-    if not _db:
-        _db = typ()
+    if _db is not None:
+        # If we have already established a connection, use that
+        return _db # type: ignore[unreachable]
+        # mypy bug: https://github.com/python/mypy/issues/5423
+
+    default_uri = "mongodb://localhost:27017"
```

Comment:
A question: any specific reason for not adding this as default value in function arguments?

Reply:
Good question!

With the code as written, it doesn't make any difference. However the intention was the following behavior:
- First time connecting:
  - Use the URI passed into this function
  - If that's empty, then use the value in the environment variable
  - If the environment variable doesn't exist and the uri is empty, then use the mongodb localhost
- After the first call:
  - If uri is empty, return the old connection
  - If the uri is non-empy, use create a new connection based on that URI

Hopefully the updated code does this. Let me know if you see any problems! Thanks for pointing this out!

Reply:
Ok, this makes sense. Thanks!